esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -5014,3 +5014,442 @@ bool CWsWorkunitsEx::onWUGetArchiveFile(IEspContext &context, IEspWUGetArchiveFi
+        publishReq->setPriority(priorityReq.str());
+    if (comment.str()) //allow empty
+        publishReq->setComment(comment.str());
+
+    if (req.getDeletePrevious())
+        publishReq->setActivate(CWUQueryActivationMode_ActivateDeletePrevious);
+    else if (req.getSuspendPrevious())
+        publishReq->setActivate(CWUQueryActivationMode_ActivateSuspendPrevious);
+    else
+        publishReq->setActivate(req.getNoActivate() ? CWUQueryActivationMode_NoActivate : CWUQueryActivationMode_Activate);
+
+    publishReq->setWait(req.getMsToWait());
+    publishReq->setNoReload(req.getNoReload());
+    publishReq->setDontCopyFiles(req.getDontCopyFiles());
+    publishReq->setAllowForeignFiles(req.getAllowForeign());
+    publishReq->setUpdateDfs(req.getUpdateDfs());
+    publishReq->setUpdateSuperFiles(req.getUpdateSuperfiles());
+    publishReq->setUpdateCloneFrom(req.getUpdateCloneFrom());
+    publishReq->setAppendCluster(!req.getDontAppendCluster());
+    publishReq->setIncludeFileErrors(true);
+
+    if (timeLimit != -1)
+        publishReq->setTimeLimit(timeLimit);
+    if (warnTimeLimit != (unsigned) -1)
+        publishReq->setWarnTimeLimit(warnTimeLimit);
+    if (!memoryLimitReq.isEmpty())
+        publishReq->setMemoryLimit(memoryLimitReq.str());
+
+    Owned<CWUPublishWorkunitResponse> publishResponse = new CWUPublishWorkunitResponse("WsWorkunits");
+    onWUPublishWorkunit(context, *publishReq, *publishResponse);
+
+    const char *id = publishResponse->getQueryId();
+    if (!isEmptyString(id))
+    {
+        const char *qs = publishResponse->getQuerySet();
+        finalResult.append("   ").append(qs ? qs : "").append('/').append(id).append(" published. ");
+    }
+    if (publishResponse->getReloadFailed())
+        finalResult.append(" Added to Queryset, but request to reload queries on cluster failed.");
+
+    const IMultiException &me = publishResponse->getExceptions();
+    if (me.ordinality())
+        readExceptionMessage(me, finalResult);
+
+    readQueryFileCopyErrors(publishResponse->getFileErrors(), finalResult);
+    addEclDefinitionActionResult(eclDefinition, finalResult.str(), "Publish", true, results);
+}
+
+void CWsWorkunitsEx::unpublishEclDefinition(IEspContext &context, const char *target, IEspWUEclDefinitionActionRequest &req,
+    IArrayOf<IConstWUEclDefinitionActionResult> &results)
+{
+    Owned<CWUQuerySetQueryActionRequest> actionReq = new CWUQuerySetQueryActionRequest("WsWorkunits");
+    actionReq->setQuerySetName(target);
+    actionReq->setAction("Delete");
+
+    IArrayOf<IEspQuerySetQueryActionItem> queries;
+    StringArray &eclDefinitions = req.getEclDefinitions();
+    for (aindex_t i = 0; i < eclDefinitions.length(); i++)
+    {
+        StringBuffer eclDefinitionName = eclDefinitions.item(i);
+        if (eclDefinitionName.trim().isEmpty())
+        {
+            addEclDefinitionActionResult(nullptr, "Empty ECL Definition name in Unpublish request", "Unpublish", true, results);
```

Comment:
Since this may get to the user, maybe here call it the query id.. instead of ECL definition.  We're just reusing the same input parameter, but they are very different.